homeassistant/components/habitica.py

```diff
@@ -0,0 +1,161 @@
+    'lvl': ST(
+        'Lvl', 'mdi:arrow-up-bold-circle-outline', 'Lvl', ["stats", "lvl"]),
+    'gp': ST('Gold', 'mdi:coin', 'Gold', ["stats", "gp"]),
+    'class': ST('Class', 'mdi:sword', '', ["stats", "class"])
+}
+
+INSTANCE_SCHEMA = vol.Schema({
+    vol.Optional(CONF_URL, default='https://habitica.com'): cv.url,
+    vol.Optional(CONF_NAME): cv.string,
+    vol.Required(CONF_API_USER): cv.string,
+    vol.Required(CONF_API_KEY): cv.string,
+    vol.Optional(CONF_SENSORS):
+        vol.All(
+            cv.ensure_list,
+            vol.Schema(vol.Unique()),
+            [vol.In(list(SENSORS_TYPES))])
+})
+
+has_unique_values = vol.Schema(vol.Unique())  # pylint: disable=invalid-name
+# because we want a handy alias
+
+
+def has_all_unique_users(value):
+    """Validate that all `api_user`s are unique."""
+    api_users = [user[CONF_API_USER] for user in value]
+    has_unique_values(api_users)
+    return value
+
+
+def has_all_unique_users_names(value):
+    """Validate that all user's names are unique and set if any is set."""
+    names = [user.get(CONF_NAME) for user in value]
+    if None in names and any(name is not None for name in names):
+        raise vol.Invalid(
+            'user names of all users must be set if any is set')
+    if not all(name is None for name in names):
+        has_unique_values(names)
+    return value
+
+
+INSTANCE_LIST_SCHEMA = vol.All(
+    cv.ensure_list,
+    has_all_unique_users,
+    has_all_unique_users_names,
+    [INSTANCE_SCHEMA])
+
+CONFIG_SCHEMA = vol.Schema({
+    DOMAIN: INSTANCE_LIST_SCHEMA
+}, extra=vol.ALLOW_EXTRA)
+
+SERVICE_API_CALL = 'api_call'
+ATTR_NAME = CONF_NAME
+ATTR_PATH = CONF_PATH
+ATTR_ARGS = "args"
+EVENT_API_CALL_SUCCESS = "{0}_{1}_{2}".format(
+    DOMAIN, SERVICE_API_CALL, "success")
+
+SERVICE_API_CALL_SCHEMA = vol.Schema({
+    vol.Required(ATTR_NAME): str,
+    vol.Required(ATTR_PATH): vol.All(cv.ensure_list, [str]),
+    vol.Optional(ATTR_ARGS): dict
+})
+
+
+async def async_setup(hass, config):
+    """Set up the habitica service."""
+    conf = config[DOMAIN]
+    data = hass.data[DOMAIN] = {}
+    websession = async_get_clientsession(hass)
+    from habitipy.aio import HabitipyAsync
+
+    class HAHabitipyAsync(HabitipyAsync):
+        """Closure API class to hold session."""
+
+        def __call__(self, **kwargs):
+            return super().__call__(websession, **kwargs)
+
+    for instance in conf:
+        url = instance[CONF_URL]
+        username = instance[CONF_API_USER]
+        password = instance[CONF_API_KEY]
+        name = instance.get(CONF_NAME)
+        config_dict = {"url": url, "login": username, "password": password}
+        api = HAHabitipyAsync(config_dict)
+        user = await api.user.get()
+        if name is None:
+            name = user['profile']['name']
+        if name in data:
+            _LOGGER.warning(
+                "Same user is specified twice: %s", name)
+        data[name] = api
+        if CONF_SENSORS in instance:
```

Comment:
Do we want this to be optional, ie without default value? Why set up an instance without any platforms? Only to be able to call the service?

Reply:
Yes. Only to be able to call service. Is this a bad use-case?

Reply:
No, I think it's ok.

Reply:
We shouldn't set up any sensors if they specify empty sensors. That would be counterintuitive. With option 1 we have both a shortcut to all possible sensors, and a way of saying no sensors. That's probably best, even though explicit is usually preferred.